code/API_definitions/location-subscriptions.yaml

```diff
@@ -0,0 +1,591 @@
+        
+        If the allocated and observed IP addresses are the same (i.e. NAT is not in use) then  the same address should be specified for both publicAddress and privateAddress.
+        
+        If NAT64 is in use, the device should be identified by its publicAddress and publicPort, or separately by its allocated IPv6 address (field ipv6Address of the Device object)
+      
+        In all cases, publicAddress must be specified, along with at least one of either privateAddress or publicPort, dependent upon which is known. In general, mobile devices cannot be identified by their public IPv4 address alone.
+
+    SingleIpv4Addr:
+      type: string
+      description: A single IPv4 address with no subnet mask
+      example: "84.125.93.10"
+      pattern: '^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])$'
+    
+    Port:
+      type: integer
+      minimum: 0
+      maximum: 65535
+    
+    Ipv6Address:
+      type: string
+      allOf:
+        - pattern: '^((:|(0?|([1-9a-f][0-9a-f]{0,3}))):)((0?|([1-9a-f][0-9a-f]{0,3})):){0,6}(:|(0?|([1-9a-f][0-9a-f]{0,3})))(\/(([0-9])|([0-9]{2})|(1[0-1][0-9])|(12[0-8])))?$'
+        - pattern: '^((([^:]+:){7}([^:]+))|((([^:]+:)*[^:]+)?::(([^:]+:)*[^:]+)?))(\/.+)?$'
+      example: "2001:db8:85a3:8d3:1319:8a2e:370:7344"
+      description: |
+        IPv6 address, following IETF 5952 format, may be specified in form <address/mask> as:
+          - address - The /128 subnet is optional for single addresses:
+            - 2001:db8:85a3:8d3:1319:8a2e:370:7344
+            - 2001:db8:85a3:8d3:1319:8a2e:370:7344/128
+          - address/mask - an IP v6 number with a mask:
+            - 2001:db8:85a3:8d3::0/64
+            - 2001:db8:85a3:8d3::/64
+    
+    ErrorInfo:
+      type: object
+      required:
+        - status
+        - code
+        - message
+      properties:
+        status:
+          type: integer
+          description: HTTP status code returned along with this error response
+        code:
+          type: string
+          description: Code given to this error
+        message:
+          type: string
+          description: Detailed error description
+    
+    Webhook:
+      type: object
+      description: Detail for event channel
+      properties:
+        notificationUrl:
+          $ref: '#/components/schemas/NotificationUrl'
+        notificationAuthToken:
+          $ref: '#/components/schemas/NotificationAuthToken'
+      required: 
+        - notificationUrl
+    
+    EventSubscriptionId:
+      type: string
+      format: uuid
+      description: EventSubscription ID in UUID format      
+    
+    SubscriptionDetail:
+      type: object
+      description: Subscription detail
+      properties:
+        device:
+          $ref: '#/components/schemas/Device'
+        area:
+          $ref: '#/components/schemas/Area'
+        eventType: 
```

Comment:
Here we may consider to convert this to an array `eventTypes` and allow subscribing to both `AREA_ENTERED` and `AREA_LEFT` for same `eventSubscriptionId, which may be a common case`. Otherwise 2 independent subscriptions are required for client interested in both. If a client is only interested in one of them, or prefer to manage both subscriptions independently, it can do it as well.

Reply:
I've slight preference to keep it single (and not array) as we have this pattern in a Device Status P/R.
I imagine having One subscription - one event type will simplify the monetization (and for developer it did not require additional work as we use same api for distinct event)

Reply:
But I see that in Device Status there is an event `ROAMING_STATUS` that covers the other two `ROAMING_ON` and `ROAMING_OFF`, so in that API it would be redundant. But here it may also common that a developer is interested in both.

Reply:
Yes - `ROAMING_ON` and `ROAMING_OFF` cover specific UC - we had a discussion in this project about very specific event.

You can have some UC with moving IoT abroad that require an update and then app are interested for `ROAMING_OFF` , while have some IoT for example that must not left national territory (and `ROAMING_ON` is subscribed as an alarm).

`ROAMING_STATUS` is the generic one for UC where both event are important.